common/common_structs/src/lib.rs

```diff
@@ -144,3 +144,17 @@ pub struct LockedAssetTokenAttributesEx<M: ManagedTypeApi> {
     pub unlock_schedule: UnlockScheduleEx<M>,
     pub is_merged: bool,
 }
+
+impl<M: ManagedTypeApi> LockedAssetTokenAttributesEx<M> {
+    pub fn average_unlock_epoch(&self) -> Epoch {
+        let mut weight_total = 0;
+        let mut weighted_sum = BigUint::<M>::zero();
+        for milestone in &self.unlock_schedule.unlock_milestones {
+            weighted_sum += milestone.unlock_percent * milestone.unlock_epoch;
+            weight_total += milestone.unlock_percent;
+        }
+
+        let weighted_average = weighted_sum / weight_total;
+        unsafe { weighted_average.to_u64().unwrap_unchecked() }
```

Comment:
Why is `unsafe` needed here?

Reply:
`unwrap_unchecked` is an unsafe function, as it leads to undefined behaviour in case of `None` (`to_u64` return `Option<u64>`), but a weighted average like this will never go over `u64`, so it's mathematically safe.